Use packit to build rpm and add it as check on pull request

filippor opened the pull request:
<!-- Include a brief summary of the changes. -->
add packit to build rpm on pull request and main branch
<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->

## Before and After Images
<!-- If your proposed changes affect the FreeCAD GUI, add before and after screenshots -->
this pr is based on pr #20996 

- trigger a rpm build on each pr targeted to main branch
- the commit on main is configured to trigger a build in a specific copr project 
 this build on main can replace the freecad-git in  pr #20996 

to enable rpm build the installation of the ["Packit-as-a-Service" GitHub application](https://github.com/marketplace/packit-as-a-service).
following https://packit.dev/docs/guide

a test pr to see this configuration in action  https://github.com/filippor/FreeCAD/pull/6
this pr is to use only packit to manage the spec file and the build removing the script used to mantain the 3 different 
build for rpm (rpkg, custom bash scrip, packit)

a sample of the copr published package on copr https://copr.fedorainfracloud.org/coprs/filippor/packit/packages/
<!--  Notes on the PR Review Process

The following section describes what the maintainers consider when reviewing your Pull Request.  These items may not require you to take any action.  This information is provided for context. Understanding what we consider will help you prepare your request for speedy approval.

You can find additional documentation about these guidelines in the [Developers handbook](https://freecad.github.io/DevelopersHandbook).

Alignment (Does the PR align with the goals and interests of the project?)
  - Does the PR have at least one issue linked, which this PR closes?
  - Has the conversation on the PR and related issue(s) reached consensus?
  - If the PR affects the GUI, is the Design Working Group (DWG) aware and have they had time to review and comment?
  - If the PR affects the GUI, did the contributor include before/after images?
  - If the PR affects standards and workflow, is the CAD Working Group (CWG) aware and have they had time to review/comment?

Impact (Does the change affect other parts of the project?)
  - Has the impact on documentation been considered and appropriate action taken?
  - Has the impact on translation been considered appropriate action taken?
  - Will the PR affect existing user documents?

Code Quality (Is code well-written and maintainable?)
  - Does the PR warrant a review by the Code Quality Working Group (CQWG)?
  - Does the change include tests?
  - Is the PR rebased on the current main branch with unnecessary commits squashed?

Release (Are there considerations related to release timing?)
  - Has the PR been considered for backporting to the latest release branch?
  - Have the release notes been considered/updated?
  -->

filippor commented:
there are same failure in ctest but all the python test pass
https://artifacts.dev.testing-farm.io/2f689048-32d8-482c-8b3b-bc3fa93b2256/
select in the top Show passed tests to see all tests executed with log

chennes commented on package/fedora/freecad.spec:
```diff
@@ -1,104 +1,69 @@
-# See FreeCAD-main/src/3rdParty/salomesmesh/CMakeLists.txt to find this out.
-%global bundled_smesh_version 7.7.1.0
+%bcond_without bundled_smesh
+# rpmbuild --without=tests:  esclude tests in %%check
```
```suggestion
# rpmbuild --without=tests:  exclude tests in %%check
```

chennes commented on package/fedora/freecad.spec:
```diff
@@ -1,104 +1,69 @@
 Group:          Applications/Engineering
-
-License:        LGPLv2+
+License:        GPL-2.0-or-later
```
```suggestion
License:        LGPL-2.0-or-later
```

chennes commented on package/fedora/freecad.spec:
```diff
@@ -1,104 +1,69 @@
-BuildRequires:  tbb-devel
+# Maintainers:  keep this list of plugins up to date
+# List plugins in %%{_libdir}/%%{name}/lib, less '.so' and 'Gui.so', here
+%global plugins AssemblyApp AssemblyGui CAMSimulator DraftUtils Fem FreeCAD Import Inspection MatGui Materials Measure Mesh MeshPart Part PartDesignGui Path PathApp PathSimulator Points QtUnitGui ReverseEngineering Robot Sketcher Spreadsheet Start Surface TechDraw Web _PartDesign area flatmesh libDriver libDriverDAT libDriverSTL libDriverUNV libE57Format libMEFISTO2 libSMDS libSMESH libSMESHDS li ...
```
The description above makes it sound like there shouldn't be the "Gui" entries here. Also, why does PartDesign appear out of order, and with an underscore?

filippor replied:
it is the library name to keep the library private and avoid other package to depend on that library. 
The gui is added in the regular expression used.
This mechanism is from the original spec file.
I'm not sure it is needed.

filippor replied:
This list is used to filter the Requires and provides with a regular expression that add ".so" and "Gui.so" to  names in the list. Removing this libraries from Provides prevent other packages to depend on that libraries (If a package try to use a library that are filtered it can't be installed because no package declared that provides the library).
The package with Gui don't have a library counterpart without trailing Gui  
_PartDesign.so is the library generated by the build. 
This list didn't change from original spec file. 
Let me know if i need to fix the list removing the trailing Gui and sort it. Or remove this part completely and use the default behaviour of rpmbuild to declare the libraries as provided.

chennes replied:
If this is just leftover from the old version of the file I don't mind leaving it, but of course feel free to clean it up if you'd like.

chennes commented on package/fedora/freecad.spec:
```diff
@@ -110,46 +75,39 @@ Requires:       %{name}-data = %{epoch}:%{version}-%{release}
-%filter_setup
-}
+%global plugin_exclude %( for i in %{plugins}; do  echo -n "\|$i\(Gui\)\?"; done )
+# idem per le librerie esportate (non hanno “Gui” opzionale)
```
Comment should be in English.

chennes commented on package/fedora/freecad.spec:
```diff
@@ -110,46 +73,39 @@ Requires:       %{name}-data = %{epoch}:%{version}-%{release}
-%filter_setup
-}
+%global plugin_exclude %( for i in %{plugins}; do  echo -n "\|$i\(Gui\)\?"; done )
+# prevent to declare Requires intenal FreeCAD libraries
```
```suggestion
# prevent declaring Requires for internal FreeCAD libraries
```

adrianinsaval commented:
we're transitioning the weekly builds to happen on the main repo so I don't know if it's a good idea to use url's from freecad-bundle for this, it'll likely be disabled in the near future, beyond that I'm lost with all the PR's we have for fedora packaging so I would like to know more of what we want from this, @filippor @ljo is your objective to be able to run rpkg yourself on your local git clone or do you just want the COPR repo working again? @filippor proposes the use of new tools for packaging is that generally desired? As I understand it this also adds rpm packaging as CI checks for PRs, seems a little excesive IMO

filippor commented:
Yes this PR add the option to build a PR as CI. In this branch Is configured to rin on demand with a comment 
/packit build and /packit test.

The url in the spec file Is ignored by packit It Is replaced with the archive configured in .packit.yaml configuration.

filippor commented:
for my pr on this subject
 #20953 is the minimal change to get copr working

#20996 use different tool to build the  package. rpkg to build the main branch each commit. custom make to build from sources in the freecad-bundle release (weekly and release)

this #21063 use packit (the latest tool proposed by fedora) to do configure builds triggered by different event configured in .packit.yaml.    Can be configured to build on each release on each commit on pull request all trigger can be configured to build automatically or manual with a comment.

in both #20996 and #21063 I fixed same issue and updated  the spec file.

filippor commented on .packit.yaml:
```diff
@@ -0,0 +1,74 @@
+    - .packit.yaml
+actions:
+  post-upstream-clone:
+    - curl -O --location --retry 3 https://raw.githubusercontent.com/FreeCAD/FreeCAD-Bundle/refs/heads/main/make_version_file.py
```
fixed verify the env var BUILD_TAG

filippor commented on .packit.yaml:
```diff
@@ -0,0 +1,74 @@
+  fmf_path: package/fedora/tests
+  notifications:
+    failure_comment:
+      message: "One of the tests failed for {commit_sha}. @admin check logs {logs_url}, packit dashboard {packit_dashboard_url} and external service dashboard {external_dashboard_url}"
```
if it is a manual check triggere by a comment I think the link to report dashboard can be helpfull but I removed the notification

filippor commented on .packit.yaml:
```diff
@@ -0,0 +1,74 @@
+  branch: only-packit
+  #replace owner with @freecad and the desired project before merging
+  owner: filippor
+  project: packit
```
For now I din't find the way to manual trigger the build or trigger dayli . I will ask on packit chat

filippor commented on package/fedora/freecad.spec:
```diff
@@ -110,46 +73,39 @@ Requires:       %{name}-data = %{epoch}:%{version}-%{release}
+    range of uses in engineering, such as architecture or other engineering
+    specialities. It is a feature-based parametric modeler with a modular software
+    architecture which makes it easy to provide additional functionality without
+    modifying the core system.
```
You are right from the specification lines with leading spaces are considered preformatted. But discover tread this as single paragraph

oursland commented:
@filippor It appears that the current state of the PR is for a daily build and push to Copr.  The title of the PR is for this to be run on each PR, which does not appear to be the case.  Could you re-title the PR?

I'm not familiar with Copr.  Does this have versioning support?  Is it possible to tie a build to a SHA within the repo, which could be helpful for issue reporting for users and bisect points for developers?

filippor commented:
> @filippor It appears that the current state of the PR is for a daily build and push to Copr. The title of the PR is for this to be run on each PR, which does not appear to be the case. Could you re-title the PR?

There Is also a manuale build for Watch PR triggered by a comment (/packit build)
> I'm not familiar with Copr. Does this have versioning support? Is it possible to tie a build to a SHA within the repo, which could be helpful for issue reporting for users and bisect points for developers?
The version information Is updated before publishing the srpm on copr

See .packit.yaml file

oursland commented:
Thanks!  It really looks like that [packit](https://packit.dev/docs/guide#github) tooling could be quite valuable!

Devs have been requesting per-PR packages for testing and this seems like a good approach, at least for Fedora.  I have a few more questions:

* Does it work for forks?
* Where do PR builds go?  Is it an artifact on the PR or is it only available on Copr?

chennes commented:
@chennes will install the Packit github application.

chennes commented:
@filippor I've installed the Packit app -- see https://github.com/packit/notifications/issues/669

filippor commented:
@chennes  if you have a fedora account you can self approve the installation with a comment on the notification

chennes commented:
@filippor I don't have a Fedora account -- should we create one that we use for this CI process?

chennes commented:
@adrianinsaval do we already maintain a team Fedora account for this sort of thing?

adrianinsaval commented:
> @adrianinsaval do we already maintain a team Fedora account for this sort of thing?

Yes I'm now on the team, if you send me your account in fedora I think I can add you too

chennes commented:
Well, I cannot activate our fedora account: I just keep getting an error during the final account creation stage (setting the password), with the very informative error message "Something went wrong while creating your account, please try again later." And of course all of their resources for getting help online require a Fedora account to sign in with first! So this is dead in the water at the moment. If anyone has any connections with the project and can help out, the account email address is fedora@freecad.org.

majamassarini commented:
> Well, I cannot activate our fedora account: I just keep getting an error during the final account creation stage (setting the password), with the very informative error message "Something went wrong while creating your account, please try again later." And of course all of their resources for getting help online require a Fedora account to sign in with first! So this is dead in the water at the moment. If anyone has any connections with the project and can help out, the account email address is [fedora@freecad.org](mailto:fedora@freecad.org).

I would suggest retrying. I wasn't able to collect any other suggestions. I hope you just encountered some issue still related to the Fedora infrastructure move.

chennes commented:
Thanks, email sent.

chennes commented:
Ok @filippor I think this is working now: can you double-check the results before merging?

filippor commented:
> Ok @filippor I think this is working now: can you double-check the results before merging?
The PR build is working. for the release The project I think will be created on copr @freecad/freecad
for the daily build you need to create an action secret named COPR_TOKEN
with the content you found in https://copr.fedorainfracloud.org/api/

filippor commented:
something is changed in the way to run tests?
https://artifacts.dev.testing-farm.io/590fc8f6-725c-426d-bcea-2203c1f9cda1/
the Test_run executable is missing 
And there is failure on
[/cmd_tests/tests/TestArch](https://artifacts.dev.testing-farm.io/590fc8f6-725c-426d-bcea-2203c1f9cda1/#work-planmjkv0gft_plan_9_default-0) test #9 on default-0, took 0 s
[/gui_tests/tests/TestAddonManagerGui](https://artifacts.dev.testing-farm.io/590fc8f6-725c-426d-bcea-2203c1f9cda1/#work-planmjkv0gft_plan_34_default-0) test #34 on default-0, took 5 s
[/gui_tests/tests/TestArch](https://artifacts.dev.testing-farm.io/590fc8f6-725c-426d-bcea-2203c1f9cda1/#work-planmjkv0gft_plan_35_default-0) test #35 on default-0, took 5 s
[/gui_tests/tests/TestArchGui](https://artifacts.dev.testing-farm.io/590fc8f6-725c-426d-bcea-2203c1f9cda1/#work-planmjkv0gft_plan_36_default-0) test #36 on default-0, took 5 s

chennes commented:
ctest should find the whole set of gtests, so all the things I listed above. If you drop me a link to the test run output I'll give it a look to make sure.

filippor commented:
[/build_time_tests/ctest](https://artifacts.dev.testing-farm.io/a327018b-c931-465a-8df5-12a70b38a2c0/#work-planhn05w19z_plan_2_default-0)test #2 on default-0, took 0

filippor commented:
Now the failing test are this ctest
The following tests FAILED:
	1369 - QuantitySpinBox_Tests_run (Subprocess aborted)

And Addon Manager
https://artifacts.dev.testing-farm.io/68858e23-8b4f-40ee-abd3-2485888fd57a/

Let me know if this error are expected and and I can skip it
